### ISSUE TITLE
How to return an error from set_psk_server_callback?

### ISSUE THREAD

**Issue by golddranks** [sfackler/rust-openssl#1061]
The `set_psk_server_callback` method is used to set the pre-shared key on a server, based on the client identity sent by the client. The callback returns a `Result<usize, ErrorStack>`. There is a problem though; if no pre-shared key is not found for the given identity, we want to return an error. However, as the return error type is `ErrorStack`, we are limited to returning errors from the OpenSSL library, not a user one. Furthermore, it seems that there is no way to generate an `ErrorStack` value. That means that I'm unable to return an error from the callback – even in an erroneous situation!

Is this an oversight in the API or is there some way to generate `ErrorStacks`?

**Comment by sfackler** [comment]
You can construct a (probably empty) ErrorStack with `ErrorStack::get` to at least get the thing compiling.

OpenSSL does provide the ability to create third-party errors that interact with its error system, and I've been meaning to expose that but haven't gotten around to it.

**Comment by golddranks** [comment]
Thanks for the super quick reply! I'll try it that way.

**Comment by sfackler** [comment]
Here's a working prototype. I think we'll eventually want a macro that can generate all of the code, but I'm not quite sure what that'll look like:

```rust
use libc::{c_char, c_int, c_ulong};
use openssl::error::Error;
use openssl_sys::ERR_STRING_DATA;
use std::marker::PhantomData;
use std::ptr;
use std::sync::Once;

#[allow(non_snake_case)]
pub const fn ERR_PACK(l: c_int, f: c_int, r: c_int) -> c_ulong {
    ((l as c_ulong & 0x0ff) << 24) | ((f as c_ulong & 0xfff) << 12) | (r as c_ulong & 0xfff)
}

#[derive(Copy, Clone)]
pub struct Function<T>(c_int, PhantomData<T>);

#[derive(Copy, Clone)]
pub struct Reason<T>(c_int, PhantomData<T>);

pub struct CustomLib(());

impl CustomLib {
    pub const MAIN: Function<CustomLib> = Function(1, PhantomData);

    pub const PHASE_OF_THE_MOON: Reason<CustomLib> = Reason(1, PhantomData);

    pub unsafe fn put(
        function: Function<CustomLib>,
        reason: Reason<CustomLib>,
        file: &'static str,
        line: u32,
    ) {
        static ONCE: Once = Once::new();
        static mut LIB: c_int = 0;

        ONCE.call_once(|| {
            openssl_sys::init();
            LIB = openssl_sys::ERR_get_next_error_library();

            static mut LIBRARY: [ERR_STRING_DATA; 2] = [
                ERR_STRING_DATA {
                    error: 0,
                    string: b"custom library\0".as_ptr() as *const c_char,
                },
                ERR_STRING_DATA {
                    error: 0,
                    string: ptr::null(),
                },
            ];
            LIBRARY[0].error = ERR_PACK(LIB, 0, 0);
            openssl_sys::ERR_load_strings(LIB, LIBRARY.as_mut_ptr());

            static mut FUNCTIONS: [ERR_STRING_DATA; 2] = [
                ERR_STRING_DATA {
                    error: ERR_PACK(0, 1, 0),
                    string: b"main\0".as_ptr() as *const c_char,
                },
                ERR_STRING_DATA {
                    error: 0,
                    string: ptr::null(),
                },
            ];
            openssl_sys::ERR_load_strings(LIB, FUNCTIONS.as_mut_ptr());

            static mut REASONS: [ERR_STRING_DATA; 2] = [
                ERR_STRING_DATA {
                    error: ERR_PACK(0, 0, 1),
                    string: b"wrong phase of the moon\0".as_ptr() as *const c_char,
                },
                ERR_STRING_DATA {
                    error: 0,
                    string: ptr::null(),
                },
            ];
            openssl_sys::ERR_load_strings(LIB, REASONS.as_mut_ptr());
        });

        openssl_sys::ERR_put_error(
            LIB,
            function.0,
            reason.0,
            file.as_ptr() as *const c_char,
            line as c_int,
        );
    }
}

macro_rules! put_custom_lib_error {
    ($function:expr, $reason:expr) => {
        unsafe { CustomLib::put($function, $reason, concat!(file!(), "\0"), line!()) }
    };
}

fn main() {
    put_custom_lib_error!(CustomLib::MAIN, CustomLib::PHASE_OF_THE_MOON);

    println!("{}", Error::get().unwrap());
}
```

Prints:

```
error:80001001:custom library:main:wrong phase of the moon:src/main.rs:94:
```

**Comment by golddranks** [comment]
Thanks for the sample code! Much appreciated.

**Comment by AviateX14** [comment]
I'm looking to do something similar with a project I'm working on, but from the client side, ideally I need a way to propgate an io:Error into ErrorStack. Using the above code works if I only want to send through a generic 'an io error occured writing the identity' but really I'd like the actual error to make it through, I'm assuming that it's not currently the most achievable thing?

This is what I'm currently working with:

```rust
builder.set_psk_client_callback(
    move | _ssl, _hint, mut identity_buffer, mut psk_buffer | {
        if let Err(_) = identity_buffer.write_all(client.identity.as_bytes()) {
            put_psk_client_error!(PskClient::CONNECT, PskClient::IO_ERROR);
            return Err(ErrorStack::get());
        }
        
        if let Err(_) = psk_buffer.write_all(&client.key) {
            put_psk_client_error!(PskClient::CONNECT, PskClient::IO_ERROR);
            return Err(ErrorStack::get());
        }

        Ok(client.key.len())
    }
);
```

**Comment by sfackler** [comment]
You're unfortunately not going to be able to stick arbitrary Rust types in the OpenSSL error stack, but you can add in an extra string of context. I'm working on a library for this on a branch - here's an example of using extra formatting:

https://github.com/sfackler/rust-openssl/blob/error-lib/openssl-errors/tests/test.rs#L44-L54

**Comment by AviateX14** [comment]
Expected as much, it’s less the arbitrary rust type and more the ability to push something more helpful in at point of call, pretty much exactly what your sample is doing. Do you have an expected timeline merging these changes into master?

**Comment by sfackler** [comment]
It's pretty much there - I just had to tweak the CI infrastructure a bit. Should be able to have it out in a day or two.

**Comment by AviateX14** [comment]
Awesome, I’ll watch out for it!

**Comment by sfackler** [comment]
Fixed by #1082.
